mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraIndexTableHandler.java

```diff
@@ -172,19 +177,21 @@ private Mono<Void> manageApplicableFlagsOnFlagsUpdate(CassandraId mailboxId, Lis
     }
 
     private Mono<Void> updateDeletedOnFlagsUpdate(CassandraId mailboxId, List<UpdatedFlags> updatedFlags) {
-        return Flux.fromIterable(updatedFlags)
-            .concatMap(flags -> updateDeletedOnFlagsUpdate(mailboxId, flags))
-            .then();
-    }
+        List<MessageUid> addDeletedUids = updatedFlags
+            .stream()
+            .filter(flag -> flag.isModifiedToSet(Flags.Flag.DELETED))
+            .map(UpdatedFlags::getUid)
+            .collect(Collectors.toList());
+        List<MessageUid> removeDeletedUids = updatedFlags
+            .stream()
+            .filter(flag -> flag.isModifiedToUnset(Flags.Flag.DELETED))
+            .map(UpdatedFlags::getUid)
+            .collect(Collectors.toList());
```

Comment:
```suggestion
        ImmutableList.Builder<MessageUid> addDeletedUidsBuilder = ImmutableList.builder();
        ImmutableList.Builder<MessageUid> removeDeletedUidsBuilder = ImmutableList.builder();

        updatedFlags
            .forEach(flag -> {
                if (flag.isModifiedToSet(Flags.Flag.DELETED)) {
                    addDeletedUidsBuilder.add(flag.getUid());
                } else if (flag.isModifiedToUnset(Flags.Flag.DELETED)) {
                    removeDeletedUidsBuilder.add(flag.getUid());
                }
            });
```
it will avoid re-stream, re-filter?